### PR TITLE
Initiate Types Rule Package and Implement MISRA RULE-6-1 and RULE-6-2

### PR DESCRIPTION
**Pull request by jeongsoolee09**
## Description

Implemented MISRA's RULE-6-1: *Bit-fields shall only be declared with an appropriate type* and MISRA's RULE-6-2: *Single-bit named bit fields shall not be of a signed type*.

## Change request type

 - [ ] Release or process automation (GitHub workflows, internal scripts)
 - [ ] Internal documentation
 - [ ] External documentation
 - [X] Query files (`.ql`, `.qll`, `.qls` or unit tests)
 - [ ] External scripts (analysis report or other code shipped as part of a release)

## Rules with added or modified queries

 - [ ] No rules added
 - [X] Queries have been added for the following rules:
     - RULE-6-1
     - RULE-6-2
 - [ ] Queries have been modified for the following rules:
     - _rule number here_

## Release change checklist

A change note ([development_handbook.md#change-notes](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#change-notes)) is required for any pull request which modifies:

 - The structure or layout of the release artifacts.
 - The evaluation performance (memory, execution time) of an existing query.
 - The results of an existing query in any circumstance.

If you are only adding new rule queries, a change note is not required.

_**Author:**_ Is a change note required? 
  - [ ] Yes
  - [X] No

🚨🚨🚨
_**Reviewer:**_ Confirm that format of *shared* queries (not the .qll file, the
.ql file that imports it) is valid by running them within VS Code. 
  - [x] Confirmed

_**Reviewer:**_ Confirm that either a change note is not required or the change note is required and has been added.
  - [x] Confirmed

## Query development review checklist

For PRs that add new queries or modify existing queries, the following checklist should be completed by both the author and reviewer:

### Author

 - [X] Have all the relevant rule package description files been checked in?
 - [X] Have you verified that the metadata properties of each new query is set appropriately?
 - [X] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [X] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [X] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [X] Does the query have an appropriate level of in-query comments/documentation?
 - [X] Have you considered/identified possible edge cases?
 - [X] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### Reviewer 

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### CODE REVIEW

**Comment by jeongsoolee09** [comment]
The suggested `cert-help-extraction.py` seems to have a bug.

```
√ codeql-coding-standards % python3.9 scripts/help/cert-help-extraction.py 
ID: ARR39-C - Converting contents at https://wiki.sei.cmu.edu/confluence/display/c/ARR39-C.+Do+not+add+or+subtract+a+scaled+integer+to+a+pointer into Markdown help file for DoNotAddOrSubtractAScaledIntegerToAPointer
ID: ARR39-C - Didn't find heading 'CERT', going to update the headings 'Noncompliant Code Example, Compliant Solution, Related Vulnerabilities, Automated Detection, Risk Assessment, Description, Related Guidelines, Bibliography'.
ID: CON30-C - Converting contents at https://wiki.sei.cmu.edu/confluence/display/c/CON30-C.+Clean+up+thread-specific+storage into Markdown help file for CleanUpThreadSpecificStorage
ID: CON30-C - Didn't find heading 'CERT', going to update the headings 'Noncompliant Code Example, Compliant Solution, Related Vulnerabilities, Automated Detection, Risk Assessment, Description'.
ID: CON31-C - Converting contents at https://wiki.sei.cmu.edu/confluence/display/c/CON31-C.+Do+not+destroy+a+mutex+while+it+is+locked into Markdown help file for DoNotAllowAMutexToGoOutOfScopeWhileLocked
ID: CON31-C - Didn't find heading 'CERT', going to update the headings 'Noncompliant Code Example, Compliant Solution, Related Vulnerabilities, Automated Detection, Risk Assessment, Description, Related Guidelines, Bibliography, CERT-CWE Mapping Notes'.
ID: CON31-C - Converting contents at https://wiki.sei.cmu.edu/confluence/display/c/CON31-C.+Do+not+destroy+a+mutex+while+it+is+locked into Markdown help file for DoNotDestroyAMutexWhileItIsLocked
ID: CON31-C - Didn't find heading 'CERT', going to update the headings 'Noncompliant Code Example, Compliant Solution, Related Vulnerabilities, Automated Detection, Risk Assessment, Description, Related Guidelines, Bibliography, CERT-CWE Mapping Notes'.
ID: CON32-C - Converting contents at https://wiki.sei.cmu.edu/confluence/display/c/CON32-C.+Prevent+data+races+when+accessing+bit-fields+from+multiple+threads into Markdown help file for PreventDataRacesWithMultipleThreads
ID: CON32-C - Didn't find heading 'CERT', going to update the headings 'Related Vulnerabilities, Compliant Solution (Bit-field, C11, Mutex), Automated Detection, Risk Assessment, Description, Compliant Solution (C11), Bibliography, Noncompliant Code Example (Bit-field)'.
ID: CON33-C - Converting contents at https://wiki.sei.cmu.edu/confluence/display/c/CON33-C.+Avoid+race+conditions+when+using+library+functions into Markdown help file for RaceConditionsWhenUsingLibraryFunctions
ID: CON33-C - Didn't find heading 'CERT', going to update the headings 'Noncompliant Code Example, Compliant Solution (Annex K, strerror_s()), Related Vulnerabilities, Automated Detection, Risk Assessment, Description, Related Guidelines, Bibliography, Compliant Solution (POSIX, strerror_r()), CERT-CWE Mapping Notes'.
ID: CON34-C - Converting contents at https://wiki.sei.cmu.edu/confluence/display/c/CON34-C.+Declare+objects+shared+between+threads+with+appropriate+storage+durations into Markdown help file for AppropriateThreadObjectStorageDurations
Traceback (most recent call last):
  File "/Users/jslee/Work/WorkRepos/codeql-coding-standards/scripts/help/cert-help-extraction.py", line 526, in <module>
    if find_heading(parsed_help, 'CERT'):
  File "/Users/jslee/Work/WorkRepos/codeql-coding-standards/scripts/shared/markdown_helpers.py", line 98, in find_heading
    if get_heading_text(h) == heading:
  File "/Users/jslee/Work/WorkRepos/codeql-coding-standards/scripts/shared/markdown_helpers.py", line 108, in get_heading_text
    return heading.children[0].children
IndexError: list index out of range
```

**Comment by lcartey** [comment]
@jeongsoolee09 Can you please create a new package for your two rules? We don't want to generate and commit files for all 30 rules in the Types package, as we are not going to implement all of them.

I would recommend calling the new package `BitfieldTypes`. You can create it like this:

 1. Update `rules.csv` so that Rule 6.1 and Rule 6.2 are assigned to the new package (`BitfieldTypes`).
 2. Generate the BitfieldTypes c package description file.
 3. Copy any changes you made for Rule 6.1 and Rule 6.2 from the existing `Types.json` to the `BitfieldTypes.json`.
 4. Generate the package files for the new package.
 5. Delete any files from the types package.

Thanks!

**Comment by jeongsoolee09** [comment]
I am changing this PR to draft to accommodate RULE-6-2 as well.

**Comment by jeongsoolee09** [comment]
Undone `Types` and made a small rule package called `BitfieldTypes`.

File: c/misra/src/rules/RULE-6-2/SingleBitNamedBitFieldsOfASignedType.ql [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+import cpp
+import codingstandards.c.misra
+
+predicate isSigned(Type type) {
+  /* Check if it's a fixed number type, because declaring fixed number types like int8_t as 1 bit is obviously absurd */
+  type instanceof FixedWidthIntegralType or 
+  /* Check if it's EXPLICITLY signed, because according to Rule 6.1, 'int' may be either signed or unsigned depending on the implementation. In the latter case, the query would lead to false positives. */
+  type instanceof IntegralType and
+  type.(IntegralType).isExplicitlySigned()
```
Comment (lcartey) [inline c/misra/src/rules/RULE-6-2/SingleBitNamedBitFieldsOfASignedType.ql:21]: Use isSigned instead.

File: c/misra/src/rules/RULE-6-2/SingleBitNamedBitFieldsOfASignedType.ql [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+ * @tags external/misra/id/rule-6-2
+ *       external/misra/obligation/required
+ */
+
+import cpp
+import codingstandards.c.misra
+
+predicate isSigned(Type type) {
+  /* Check if it's a fixed number type, because declaring fixed number types like int8_t as 1 bit is obviously absurd */
+  type instanceof FixedWidthIntegralType or
```
Comment (lcartey) [inline c/misra/src/rules/RULE-6-2/SingleBitNamedBitFieldsOfASignedType.ql:18]: I don't think we need this condition.

File: c/misra/src/rules/RULE-6-2/SingleBitNamedBitFieldsOfASignedType.ql [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+ * @precision very-high
+ * @problem.severity error
+ * @tags external/misra/id/rule-6-2
+ *       external/misra/obligation/required
+ */
+
+import cpp
+import codingstandards.c.misra
+
+predicate isSigned(Type type) {
```
Comment (lcartey) [inline c/misra/src/rules/RULE-6-2/SingleBitNamedBitFieldsOfASignedType.ql:16]: I think you will be able to inline this predicate once you simplify it.

File: c/misra/src/rules/RULE-6-2/SingleBitNamedBitFieldsOfASignedType.ql [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+}
+
+/* Check if the DECLARED bit-fields is a single bit, because Rule 6.2 also intends to catch confusion on the programmers' part. Consider:
+
+struct S {
+  int32_t x: 1;
+}
+
+In this case, field x is essentially of 32 bits, but is declared as 1 bit and its type int32_t is signed. Therefore, it indicates confusion by the programmer, which is exactly what this rule intends to find. */
+predicate isSingleBit(BitField bitField) {
```
Comment (lcartey) [inline c/misra/src/rules/RULE-6-2/SingleBitNamedBitFieldsOfASignedType.ql:31]: I think you can inline this predicate

**Review by lcartey — CHANGES_REQUESTED** [review]
Looks good, just a couple of changes.

File: rule_packages/c/Types.json [hunk 5]
```diff
@@ -0,0 +1,482 @@
+{
```
Comment (lcartey) [inline rule_packages/c/Types.json:1]: Delete this file, regenerate the `BitfieldTypes` package, and delete the excess tests.

**Comment by jeongsoolee09** [comment]
We can ignore implementation-specific integer types: [GCC](https://gcc.gnu.org/onlinedocs/gcc-8.5.0/gcc/C-Implementation.html#C-Implementation) and [Clang](https://releases.llvm.org/10.0.0/tools/clang/docs/LanguageExtensions.html) does not define any additional types that are equivalent to `int`.

File: c/misra/src/rules/RULE-6-1/BitFieldsShallOnlyBeDeclaredWithAnAppropriateType.ql [hunk 6]
```diff
@@ -13,26 +13,24 @@
...
-not isExcluded(bitField, BitfieldTypesPackage::bitFieldsShallOnlyBeDeclaredWithAnAppropriateTypeQuery()) and
-/* A violation would neither an appropriate primitive type nor an appropriate typedef. */
-not (isAppropriatePrimitive(bitField.getType()) or isAppropriateTypedef(bitField.getType()))
-select bitField, "Bit-field " + bitField + " is declared on type " + bitField + "."
\ No newline at end of file
+  not isExcluded(bitField,
+    BitfieldTypesPackage::bitFieldsShallOnlyBeDeclaredWithAnAppropriateTypeQuery()) and
+  /* A violation would neither an appropriate primitive type nor an appropriate typedef. */
+  not (
+    isAppropriatePrimitive(bitField.getType()) or
```
[the remark this answers is not in the document]
Comment (jeongsoolee09) [reply c/misra/src/rules/RULE-6-1/BitFieldsShallOnlyBeDeclaredWithAnAppropriateType.ql]: Didn't know that resolveTypedefs is an identity operation on types that have no aliases.

**Review by lcartey — APPROVED** [review]

File: c/misra/test/rules/RULE-6-2/test.c [hunk 7]
```diff
@@ -0,0 +1,17 @@
+#include <stdint.h>
+
+struct SampleStruct {
+  int x1 : 1; // NON_COMPILANT: very likely be signed, but if it's not, the
```
Comment (rvermeulen) [inline c/misra/test/rules/RULE-6-2/test.c:4]: NON_COMPLIANT

**Pull request merged**